lib/ecto/repo/model.ex

```diff
@@ -54,16 +57,23 @@ defmodule Ecto.Repo.Model do
 
     with_transactions_if_callbacks repo, adapter, model, opts,
                                    ~w(before_update after_update)a, fn ->
+      # Callbacks.__apply__/3 "ignores" undefined callbacks and simply returns
+      # the changeset unchanged in case the callback is missing.
       changeset = Callbacks.__apply__(model, :before_update, changeset)
       changes   = validate_changes(:update, model, fields, changeset)
 
-      pk_filter = Planner.fields(:update, model, pk_filter(model, struct))
+      filters = Planner.fields(:update, model, filters!(model, struct, changeset))
 
       if changes == [] do
-        changes = pk_filter
```

Comment:
This should be only the pk_filter. Basically, we only set this if you are doing an update with no changes, so we basically set the ID equal to the current ID so we can at least send something to the database.